site-src/geps/gep-735.md

```diff
@@ -0,0 +1,202 @@
+# GEP-735: TCP and UDP addresses and port matching
+
+* Issue: [#735](https://github.com/kubernetes-sigs/gateway-api/issues/735)
+* Status: Provisional
+
+## TLDR
+
+Spec for matching source and destination addresses & ports on L4 APIs.
+
+## Goals
+
+- add matching rules for address/port to `TCPRoute`
+- add matching rules for address/port to `UDPRoute`
+- intentionally avoid type definitions that would make it hard to expand later
+
+## Non-Goals
+
+- defining match rules for source ports
+
+## Introduction
+
+While `TCPRoute` and `UDPRoute` currently support custom matching extensions,
+there is desire among the community to include some "fundamental" matching
+options in the spec that cover the most common requirements. In this GEP we
+request address and port matching for these APIs in order to support a standard
+for some of the commonplace setups of gateway implementations. Matching is
+intended to be covered for both _source_ and _destination_ to enable a finer
+level of tuning options for L4 traffic routing at a level below the `Gateway`.
+
+When this GEP was started the intention was to cover source port matching but
+the use cases for such were significantly less obvious than matching for other
+types, and it seemed likely that a port _range_ would be wanted in that case.
+Given the lack of strong requirements source ports were left out for this
+iteration with the intention that it can be very easily added later to the
+`TrafficMatches` type there's stronger requirements established.
+
+## API
+
+The API changes include the following new types:
+
+- `NetworkMatch` to indicate the IP for address matching
+- `TrafficMatches` to configure matching according to network address and ports
+
+These types enable the port and address matching required, with some
+active considerations about how to leave these open ended for later expansion.
+
+### NetworkMatch Type
+
+A new `NetworkMatch` type provides the targeting mechanism for match inclusion
+of a given network address:
+
+```go
+type NetworkMatch struct {
+	// Address is an IP address which refers to either an IPv4 or IPv6 address.
+	//
+	// Examples: `127.0.0.1`, `::1`
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=45
+	Address string `json:"address"`
```

Comment:
Just want to make sure I'm clear: are you suggesting:

a) type for IP addresses (v4, v6)
b) type for address (e.g. can include hostnames)
c) both

Reply:
I think the idea of a "NamedAddress" here makes sense, similar to the address type we have on Gateway. I don't think we'd need to distinguish between v4 and v6 addresses though, but don't feel too strongly about that.

Reply:
let me know your thoughts f82c90bf0b696ec7ca333a2625e34d17f8a5d472

Reply:
1462088fdeada16de4ed7f624f465b5aa0d103e0 changes this to support both `IPAddress` and `NamedAddress` options, let me know your thoughts :bow:

Reply:
NamedAddress feels a bit like ImplementationSpecific in path match (which we got rid off).
@robscott @howardjohn Does NamedAddress by itself (without the additional context of the provider) enough here?

Reply:
That's a good point, NamedAddress is going to have a different meaning to most implementations. That's easier to solve on Gateways because each will only be implemented by a single controller. This feels similar in nature to RegEx path matching - each implementation will have a different take on it, but it means something conceptually similar. 

In most cases you would not want to specify a RegEx match on a Route implemented by multiple controllers unless you knew they supported the same variation of RegEx. I think this would come with the same constraints. Although that's not ideal, I do think it's a sufficiently valuable concept to include. We likely need to document the limitations of these implementation-specific types though.

Reply:
>NamedAddress feels a bit like ImplementationSpecific in path match

@hbagdi I think there's a significant difference here. One of the key problems with "ImplementationSpecific" path matching was that any given implementation may have multiple extended kinds of path matching they'd like to support, and "ImplementationSpecific" could only really support one per implementation. With NamedAddress, I think there will be at most one interpretation per implementation. We already have a "NamedAddress" concept in the API that's fairly well defined, so it's not so much a question of if we should add the concept of a "NamedAddress", but if it makes sense here. 

> given the contention I'm leaning towards removing the NamedAddress for now 

@shaneutt my personal preference is that we leave this in place. I think it's a natural extension point and helps show the purpose of even having an address type here. Although it is something that will have a different meaning per implementation, that matches how NamedAddress already works elsewhere in the API, and seems conceptually similar to RegEx matching.